barretenberg/cpp/src/barretenberg/avm_fuzzer/harness/internal_call.fuzzer.cpp

```diff
@@ -0,0 +1,331 @@
+        std::uniform_int_distribution<size_t> index_dist(0, num_events - 1);
+        size_t value_idx = index_dist(rng);
+        // TODO(MW): gate by max - num_events so we don't hit the overflow case?
+        std::uniform_int_distribution<uint32_t> pc_dist(0, std::numeric_limits<uint32_t>::max());
+        input.local_pcs[value_idx] = pc_dist(rng);
+        break;
+    }
+    case 4: {
+        // Toggle testing error case where we try to pop off an empty stack (just for gadget coverage):
+        input.extra_pop = !input.extra_pop;
+        break;
+    }
+    default:
+        break;
+    }
+    // Serialize mutated input back to buffer
+    input.to_buffer(data);
+
+    if (max_size > sizeof(InternalCallFuzzerInput)) {
+        return sizeof(InternalCallFuzzerInput);
+    }
+
+    return sizeof(InternalCallFuzzerInput);
+}
+
+// NOTE: context->serialize_context_event() causes stack overflow :(
+ContextEvent fill_context_event(std::unique_ptr<ContextInterface>& context,
+                                InternalCallStackManagerInterface& internal_call_stack_manager)
+{
+    return {
+        .id = context->get_context_id(),
+        .pc = context->get_pc(),
+        .internal_call_id = internal_call_stack_manager.get_call_id(),
+        .internal_call_return_id = internal_call_stack_manager.get_return_call_id(),
+        .next_internal_call_id = internal_call_stack_manager.get_next_call_id(),
+    };
+}
+
+void fuzz_internal_call(std::vector<ExecutionEvent>& ex_events,
+                        std::unique_ptr<ContextInterface>& context,
+                        InternalCallStackManagerInterface& internal_call_stack_manager,
+                        uint32_t loc)
+{
+    ExecutionEvent ex_event = { .wire_instruction = call_instr,
+                                .before_context_event = fill_context_event(context, internal_call_stack_manager) };
+    // Execution.execute pre-dispatch:
+    context->set_next_pc(context->get_pc() + static_cast<uint32_t>(call_instr.size_in_bytes()));
+
+    // Execution.internal_call(context, loc) - internal_call_stack_manager.push() emits the internal call stack
+    // event:
+    internal_call_stack_manager.push(context->get_pc(), context->get_next_pc());
+    context->set_next_pc(loc);
+
+    // Execution.execute post-dispatch:
+    context->set_pc(context->get_next_pc());
+
+    ex_event.after_context_event = fill_context_event(context, internal_call_stack_manager);
+    ex_events.push_back(ex_event);
+}
+
+void fuzz_internal_return(std::vector<ExecutionEvent>& ex_events,
+                          std::unique_ptr<ContextInterface>& context,
+                          InternalCallStackManagerInterface& internal_call_stack_manager)
+{
+    // TODO(MW): Case where we do not have a return i.e other exit condition?
+    ExecutionEvent ex_event = { .wire_instruction = ret_instr,
+                                .before_context_event = fill_context_event(context, internal_call_stack_manager) };
```

Comment:
Should this fuzzer cover other exit conditions from an internal call? That might be a little too involved and doesn't really test internal calls specifically, more what the following exit rows would do